Feature/json validation

kmcelwee opened the pull request:
Seems to be working!

kmcelwee committed:
Add command line JSON validation

kmcelwee committed:
Add goodtables output to artifacts

kmcelwee committed:
Send message if the schema is invalid

kmcelwee committed:
Add special message formatting and update environment names and logic

kmcelwee committed:
Remove second webhook

kmcelwee committed:
Merge branch 'master' into feature/json-validation

rlskoeser commented on .github/goodtables_report.py:
```diff
@@ -43,8 +46,43 @@
 ERROR_FILE = os.environ['ERROR_FILE']
```
On re-review I was confused by this variable name. Is it accurate to call it an error file? Is it not just validation output?

rlskoeser commented on .github/goodtables_report.py:
```diff
@@ -43,8 +46,43 @@
+        }
+
+        # post stringified payload to url
+        requests.post(os.environ['SLACK_GOODTABLES_WEBHOOK'], json.dumps(j_payload))
```
Would be a little cleaner and simpler to get the environment variables once like you do the other global variables. Could even make a list of slack webhooks to simplify the logic where you're checking and posting twice — although I think you only do it twice, so not a huge gain.

rlskoeser commented on .github/goodtables_report.py:
```diff
@@ -43,8 +46,43 @@
+
 
 def main():
+    if is_invalid_schema():
```
Do we have to check this every time? Isn't there something you can key on in the validation output to determine if validation actually ran?

rlskoeser reviewed (commented):
Looks reasonable. A couple comments that could improve it, but all pretty minor. 

I would prefer if there's a way not to check if the schema is invalid every time. You have to open the goodtable output no matter what, is there anything we can key on to determine if the schema check is needed? When I looked at it before, it looked like the output didn't include any table information when the schema was invalid.

Not in this change set, but I didn't notice before — where you're generating the `stdout_string`, it would be more readable and more pythonic to make a list of strings and then join them with newlines  (I think that's still more efficient in current versions of python, although probably not a real factor in this case).

kmcelwee committed:
Change env name from ERROR_FILE to VALIDATION_OUTPUT_FILE

kmcelwee committed:
Create env variables to reduce repetition, and use variables we already created

kmcelwee committed:
Stop early if tables property isn't empty

kmcelwee committed:
Merge branch 'feature/json-validation' of github.com:Princeton-CDH/pemm-data into feature/json-validation

kmcelwee commented:
I think I addressed everything, but I'm not sure if how I addressed https://github.com/Princeton-CDH/pemm-data/pull/12/commits/6ea975b3c0a718fea69a9b674955dae885ce0324 is what you wanted

rlskoeser commented on .github/goodtables_report.py:
```diff
@@ -62,30 +68,32 @@ def is_invalid_schema():
+    with open(VALIDATION_OUTPUT_FILE) as f: 
         error_json = json.load(f)
 
+    # Check if schema is valid
+    if is_invalid_schema(error_json):
+        return
+
```
Took me a minute reading this to see how you changed it (I think because it wasn't quite what I expected). It doesn't seem logical to me for the invalid schema check to depend on or even know about the validation output (the schema is valid or not independent of that; it's just that we only care about it in one of those cases). 

You have the basic logic right  — just move the error json table check out of the method and flip it so you only call the `is_invalid_schema`  check when they aren't present.

kmcelwee replied:
Great, I'll move it!

rlskoeser reviewed (approved):
All the other updates look good!

kmcelwee committed:
Fix syntax error

kmcelwee committed:
Move logic outside is_invalid_schema function

kmcelwee committed:
Remove error max and add logic to make it optional

kmcelwee committed:
Fix syntax error

rlskoeser reviewed (approved):
Revisions look good to me!
